File: src/mergeLine.js
/**
 * The merge-line traversal of the Lee & Wong divide-and-conquer merge. Walks a
 * bisector between two merged halves, hopping across the nearest intersecting
 * bisector on either side until the merge line exits the canvas (or an orphaned
 * bisector needs to be unwound).
 */

import {angle, samePoint} from './geometry.js';
import {bisectorIntersection, trimBisector, findHopTo, isBisectorTrapped, getExtremePoint, removeBisector} from './bisector.js';

/**
 * Determine which border is crossed first: "right", "left", or null when both
 * are equidistant.
 *
 * @param {{bisector: Bisector, point: Array}} cropR
 * @param {{bisector: Bisector, point: Array}} cropL
 * @param {Array} currentCropPoint - [x,y]
 * @returns {string|null}
 */
export function determineFirstBorderCross(cropR, cropL, currentCropPoint){
    if(Math.abs(cropR.point[1] - currentCropPoint[1]) === Math.abs(cropL.point[1] - currentCropPoint[1])){
        return null;
    }
    else{
        return Math.abs(cropR.point[1] - currentCropPoint[1]) < Math.abs(cropL.point[1] - currentCropPoint[1]) ? "right" : "left";
    }
}

/**
 * Candidates on one side of the current merge line: the side-site bisectors
 * that intersect the merge line, sorted by the angle of their far site.
 *
 * @param {Site} sideSite
 * @param {Site} otherSite
 * @param {Bisector} currentBisector
 * @param {Array} currentCropPoint - [x,y]
 * @param {Bisector} crossedBorder
 * @param {boolean} goUp
 * @param {Metric} metric
 * @param {boolean} ascending - sort direction (left side sorts descending)
 * @returns {Array<{bisector: Bisector, point: Array}>}
 */
function cropCandidates(sideSite, otherSite, currentBisector, currentCropPoint, crossedBorder, goUp, metric, ascending){
    return sideSite.bisectors
        .map(e => {return {bisector: e, point: bisectorIntersection(currentBisector, e)}})
        .filter(e => {
            let hopTo = findHopTo(e.bisector, sideSite);
            return e.point && (goUp === metric.isUpward(hopTo, sideSite, otherSite, goUp)) && (!samePoint(e.point, currentCropPoint) || e.bisector !== crossedBorder);
        })
        .sort((a, b) => {
            let angleA = angle(sideSite.site, findHopTo(a.bisector, sideSite).site);
            let angleB = angle(sideSite.site, findHopTo(b.bisector, sideSite).site);
            return ascending ? angleA - angleB : angleB - angleA;
        })
        .filter((e, i, candidates) => {
            let hopTo = findHopTo(e.bisector, sideSite);
            let newMergeLine = metric.bisector(otherSite, hopTo);
            trimBisector(newMergeLine, e.bisector, e.point, metric);
            return candidates.every(d => !isBisectorTrapped(findHopTo(d.bisector, sideSite), newMergeLine, metric) || findHopTo(d.bisector, sideSite) === hopTo);
        });
}

function noCrop(goUp){
    return {bisector: null, point: goUp ? [Infinity, Infinity] : [-Infinity, -Infinity]};
}

/**
 * Walk the merge line between two partially-merged diagrams, recording each
 * bisector segment of the merge line in mergeArray.
 *
 * @param {Site} currentR
 * @param {Site} currentL
 * @param {Bisector} currentBisector
 * @param {Array} currentCropPoint - [x,y]
 * @param {boolean} goUp
 * @param {Metric} metric
 * @param {Bisector} crossedBorder
 * @param {Array} mergeArray
 * @returns {Array<Bisector>}
 */
export function walkMergeLine(currentR, currentL, currentBisector, currentCropPoint, goUp, metric, crossedBorder = null, mergeArray = []){

    while(true){

        if(
            !currentBisector.sites.every(e => e === currentR || e === currentL)
        ){
            currentBisector = metric.bisector(currentR, currentL);
            trimBisector(currentBisector, crossedBorder, currentCropPoint, metric);
            mergeArray.push(currentBisector);
        }

        let cropLArray = cropCandidates(currentL, currentR, currentBisector, currentCropPoint, crossedBorder, goUp, metric, false);
        let cropRArray = cropCandidates(currentR, currentL, currentBisector, currentCropPoint, crossedBorder, goUp, metric, true);

        let cropL = cropLArray.length > 0 ? cropLArray[0] : noCrop(goUp);
        let cropR = cropRArray.length > 0 ? cropRArray[0] : noCrop(goUp);

        // If no intersection, the merge line is finished.
        if(
            !cropL.bisector && !cropR.bisector
        ){
            // Check for orphaned bisectors on either side.
            let leftOrphan = checkForOrphans(currentR, currentL, goUp, metric);
            let rightOrphan = checkForOrphans(currentL, currentR, goUp, metric);

            if(leftOrphan){
                removeBisector(leftOrphan);
                let hopTo = findHopTo(leftOrphan, currentL);
                currentR = findCorrectW(currentR, hopTo, metric);
                let newMergeBisector = metric.bisector(hopTo, currentR);
                mergeArray.push(newMergeBisector);
                currentBisector = newMergeBisector;
                currentL = hopTo;
                continue;
            }
            else if(rightOrphan){
                removeBisector(rightOrphan);
                let hopTo = findHopTo(rightOrphan, currentR);
                currentL = findCorrectW(currentL, hopTo, metric);
                let newMergeBisector = metric.bisector(hopTo, currentL);
                mergeArray.push(newMergeBisector);
                currentBisector = newMergeBisector;
                currentR = hopTo;
                continue;
            }

            return mergeArray;
        }

        // Cross the nearest intersecting bisector (or both when equidistant).
        if(determineFirstBorderCross(cropR, cropL, currentCropPoint) === "right"){
            trimBisector(cropR.bisector, currentBisector, cropR.point, metric);
            trimBisector(currentBisector, cropR.bisector, cropR.point, metric);
            currentBisector.intersections.push(cropR.point);
            crossedBorder = cropR.bisector;
            currentR = findHopTo(cropR.bisector, currentR);
            currentCropPoint = cropR.point;
        }
        else if(determineFirstBorderCross(cropR, cropL, currentCropPoint) === "left"){
            trimBisector(cropL.bisector, currentBisector, cropL.point, metric);
            trimBisector(currentBisector, cropL.bisector, cropL.point, metric);
            currentBisector.intersections.push(cropL.point);
            crossedBorder = cropL.bisector;
            currentL = findHopTo(cropL.bisector, currentL);
            currentCropPoint = cropL.point;
        }
        else{
            if(cropR.bisector){
                trimBisector(cropR.bisector, currentBisector, cropR.point, metric);
                trimBisector(currentBisector, cropR.bisector, cropR.point, metric);
                currentBisector.intersections.push(cropR.point);
                crossedBorder = cropR.bisector;
                currentR = findHopTo(cropR.bisector, currentR);
                currentCropPoint = cropR.point;
            }
            if(cropL.bisector){
                trimBisector(cropL.bisector, currentBisector, cropL.point, metric);
                trimBisector(currentBisector, cropL.bisector, cropL.point, metric);
                currentBisector.intersections.push(cropL.point);
                crossedBorder = cropL.bisector;
                currentL = findHopTo(cropL.bisector, currentL);
                currentCropPoint = cropL.point;
            }
        }
    }
}

/**
 * Determine the starting bisector for the merge process.
 *
 * @param {Site} w - starting site
 * @param {Site} nearestNeighbor
 * @param {Metric} metric
 * @param {Array} lastIntersect - [x,y]
 * @returns {{startingBisector: Bisector, w: Site, nearestNeighbor: Site, startingIntersection: Array}}
 */
export function determineStartingBisector(w, nearestNeighbor, metric, lastIntersect = null){

    let z = [metric.width, w.site[1]];

    if(!lastIntersect){
        lastIntersect = w.site;
    }

    let zline = {points: [w.site, z]};

    let intersection = nearestNeighbor.bisectors.map(bisector => {
        return {point: bisectorIntersection(zline, bisector), bisector: bisector}
    }).find(intersection => intersection.point);

    if(intersection && metric.distance(w.site, intersection.point) > metric.distance(nearestNeighbor.site, intersection.point)){
        var startingBisector = metric.bisector(w, nearestNeighbor);
        return {
            startingBisector: startingBisector,
            w: w,
            nearestNeighbor: nearestNeighbor,
            startingIntersection: intersection.point ? intersection.point : w.site
        };
    }
    else if(intersection && metric.distance(w.site, intersection.point) < metric.distance(nearestNeighbor.site, intersection.point) && intersection.point[0] > lastIntersect[0]){
        let nextR = findHopTo(intersection.bisector, nearestNeighbor);
        return determineStartingBisector(w, nextR, metric, intersection.point);
    }
    else{
        w = findCorrectW(w, nearestNeighbor, metric);

        let startingBisector = metric.bisector(w, nearestNeighbor);

        return {
            startingBisector: startingBisector,
            w: w,
            nearestNeighbor: nearestNeighbor,
            startingIntersection: intersection ? intersection.point : w.site
        };
    }
}

/**
 * Ensure the starting point does not produce a trapped bisector.
 *
 * @param {Site} w
 * @param {Site} nearestNeighbor
 * @param {Metric} metric
 * @returns {Site}
 */
export function findCorrectW(w, nearestNeighbor, metric){

    var startingBisector = metric.bisector(w, nearestNeighbor);

    let wTrap = w.bisectors.map(e => {
        let hopTo = findHopTo(e, w);
        return {hopTo: hopTo, isTrapped: isBisectorTrapped(hopTo, startingBisector, metric)}
    })
    .filter(e => e.isTrapped)
    .sort((a,b) => metric.distance(a.hopTo.site, nearestNeighbor.site) - metric.distance(b.hopTo.site, nearestNeighbor.site))[0];

    if(wTrap){
        return findCorrectW(wTrap.hopTo, nearestNeighbor, metric);
    }
    else{
        return w;
    }
}

/**
 * Recursively find an orphaned bisector on the trapped side of a merge.
 *
 * @param {Site} trapper
 * @param {Site} trapped
 * @param {boolean} goUp
 * @param {Metric} metric
 * @returns {Bisector|null}
 */
export function checkForOrphans(trapper, trapped, goUp, metric){

    let orphan = trapped.bisectors.filter(bisector => {
        let hopTo = findHopTo(bisector, trapped);
        return goUp === hopTo.site[1] < trapped.site[1] && isBisectorTrapped(trapper, bisector, metric);
    }).sort((a,b) => {

        let hopToA = findHopTo(a, trapped);
        let hopToB = findHopTo(b, trapped);

        let mergeLineA = metric.bisector(hopToA, trapper);
        let mergeLineB = metric.bisector(hopToB, trapper);

        let extremeA = getExtremePoint(mergeLineA, goUp);
        let extremeB = getExtremePoint(mergeLineB, goUp);

        return goUp ? extremeB - extremeA : extremeA - extremeB;
    })[0];

    return orphan ? orphan : null;
}

File: py_ai/merge_line.py
"""
The merge-line traversal of the Lee & Wong divide-and-conquer merge. Walks a
bisector between two merged halves, hopping across the nearest intersecting
bisector on either side until the merge line exits the canvas (or an orphaned
bisector needs to be unwound).
"""

from geometry import angle, same_point
from bisector import (
    bisector_intersection,
    trim_bisector,
    find_hop_to,
    is_bisector_trapped,
    get_extreme_point,
    remove_bisector,
)


def determine_first_border_cross(crop_r, crop_l, current_crop_point):
    """Determine which border to cross first"""
    if abs(crop_r["point"][1] - current_crop_point[1]) == abs(
        crop_l["point"][1] - current_crop_point[1]
    ):
        return None
    elif abs(crop_r["point"][1] - current_crop_point[1]) < abs(
        crop_l["point"][1] - current_crop_point[1]
    ):
        return "right"
    else:
        return "left"


def walk_merge_line(
    current_r,
    current_l,
    current_bisector,
    current_crop_point,
    go_up,
    metric,
    crossed_bisector=None,
    merge_array=None,
):
    """
    Walk along merge line to find intersections

    :param current_r: Site
    :param current_l: Site
    :param current_bisector: Bisector
    :param current_crop_point: [x, y]
    :param go_up: boolean
    :param metric: Metric strategy object
    :param crossed_bisector: Bisector or None
    :param merge_array: list of Bisectors
    :returns: list of Bisectors
    """
    if merge_array is None:
        merge_array = []

    # Ensure both sites are in current bisector
    if not all(e == current_r or e == current_l for e in current_bisector["sites"]):
        current_bisector = metric["bisector"](current_r, current_l)
        trim_bisector(current_bisector, crossed_bisector, current_crop_point, metric)
        merge_array.append(current_bisector)

    # Process left bisectors
    crop_l_array = []
    for e in current_l["bisectors"]:
        point = bisector_intersection(current_bisector, e)
        if point:
            hop_to = next((d for d in e["sites"] if d != current_l), None)
            if hop_to and go_up == metric["is_upward"](
                hop_to, current_l, current_r, go_up
            ):
                if not same_point(point, current_crop_point) or e != crossed_bisector:
                    crop_l_array.append({"bisector": e, "point": point})

    # Sort by angle
    crop_l_array.sort(
        key=lambda x: angle(
            current_l["site"], find_hop_to(x["bisector"], current_l)["site"]
        )
    )

    # Filter trapped bisectors
    filtered_l = []
    for i, e in enumerate(crop_l_array):
        hop_to = find_hop_to(e["bisector"], current_l)
        new_merge_line = metric["bisector"](current_r, hop_to)
        trim_bisector(new_merge_line, e["bisector"], e["point"], metric)

        # Check if trapped
        is_trapped = all(
            not is_bisector_trapped(
                find_hop_to(d["bisector"], current_l), new_merge_line, metric
            )
            or find_hop_to(d["bisector"], current_l) == hop_to
            for d in crop_l_array[:i]
        )
        if is_trapped:
            filtered_l.append(e)
    crop_l_array = filtered_l

    # Process right bisectors
    crop_r_array = []
    for e in current_r["bisectors"]:
        point = bisector_intersection(current_bisector, e)
        if point:
            hop_to = next((d for d in e["sites"] if d != current_r), None)
            if hop_to and go_up == metric["is_upward"](
                hop_to, current_r, current_l, go_up
            ):
                if not same_point(point, current_crop_point) or e != crossed_bisector:
                    crop_r_array.append({"bisector": e, "point": point})

    crop_r_array.sort(
        key=lambda x: angle(
            current_r["site"], find_hop_to(x["bisector"], current_r)["site"]
        )
    )

    filtered_r = []
    for i, e in enumerate(crop_r_array):
        hop_to = find_hop_to(e["bisector"], current_r)
        new_merge_line = metric["bisector"](current_l, hop_to)
        trim_bisector(new_merge_line, e["bisector"], e["point"], metric)

        is_trapped = all(
            not is_bisector_trapped(
                find_hop_to(d["bisector"], current_r), new_merge_line, metric
            )
            or find_hop_to(d["bisector"], current_r) == hop_to
            for d in crop_r_array[:i]
        )
        if is_trapped:
            filtered_r.append(e)
    crop_r_array = filtered_r

    # Determine crop points
    infinity_pt = (
        [float("inf"), float("inf")] if go_up else [-float("inf"), -float("inf")]
    )
    crop_l = (
        crop_l_array[0]
        if crop_l_array and crop_l_array[0] != current_bisector
        else {"bisector": None, "point": infinity_pt}
    )
    crop_r = (
        crop_r_array[0]
        if crop_r_array and crop_r_array[0] != current_bisector
        else {"bisector": None, "point": infinity_pt}
    )

    # If no intersection, we're done
    if not crop_l["bisector"] and not crop_r["bisector"]:
        # Check for orphans
        left_orphan = check_for_orphans(current_r, current_l, go_up, metric)
        right_orphan = check_for_orphans(current_l, current_r, go_up, metric)

        if left_orphan:
            remove_bisector(left_orphan)

            hop_to = find_hop_to(left_orphan, current_l)
            current_r = find_correct_w(current_r, hop_to, metric)
            new_merge_bisector = metric["bisector"](hop_to, current_r)
            merge_array.append(new_merge_bisector)

            return walk_merge_line(
                current_r,
                hop_to,
                new_merge_bisector,
                current_crop_point,
                go_up,
                metric,
                crossed_bisector,
                merge_array,
            )
        elif right_orphan:
            remove_bisector(right_orphan)

            hop_to = find_hop_to(right_orphan, current_r)
            current_l = find_correct_w(current_l, hop_to, metric)
            new_merge_bisector = metric["bisector"](hop_to, current_l)
            merge_array.append(new_merge_bisector)

            return walk_merge_line(
                hop_to,
                current_l,
                new_merge_bisector,
                current_crop_point,
                go_up,
                metric,
                crossed_bisector,
                merge_array,
            )

        return merge_array

    # Determine which point to cross first
    cross = determine_first_border_cross(crop_r, crop_l, current_crop_point)

    if cross == "right":
        trim_bisector(crop_r["bisector"], current_bisector, crop_r["point"], metric)
        trim_bisector(current_bisector, crop_r["bisector"], crop_r["point"], metric)
        current_bisector["intersections"].append(crop_r["point"])
        crossed_bisector = crop_r["bisector"]
        current_r = next(
            (s for s in crop_r["bisector"]["sites"] if s != current_r), current_r
        )
        current_crop_point = crop_r["point"]
    elif cross == "left":
        trim_bisector(crop_l["bisector"], current_bisector, crop_l["point"], metric)
        trim_bisector(current_bisector, crop_l["bisector"], crop_l["point"], metric)
        current_bisector["intersections"].append(crop_l["point"])
        crossed_bisector = crop_l["bisector"]
        current_l = next(
            (s for s in crop_l["bisector"]["sites"] if s != current_l), current_l
        )
        current_crop_point = crop_l["point"]
    else:
        # Both
        if crop_r["bisector"]:
            trim_bisector(crop_r["bisector"], current_bisector, crop_r["point"], metric)
            trim_bisector(current_bisector, crop_r["bisector"], crop_r["point"], metric)
            current_bisector["intersections"].append(crop_r["point"])
            crossed_bisector = crop_r["bisector"]
            current_r = next(
                (s for s in crop_r["bisector"]["sites"] if s != current_r), current_r
            )
            current_crop_point = crop_r["point"]

        if crop_l["bisector"]:
            trim_bisector(crop_l["bisector"], current_bisector, crop_l["point"], metric)
            trim_bisector(current_bisector, crop_l["bisector"], crop_l["point"], metric)
            current_bisector["intersections"].append(crop_l["point"])
            crossed_bisector = crop_l["bisector"]
            current_l = next(
                (s for s in crop_l["bisector"]["sites"] if s != current_l), current_l
            )
            current_crop_point = crop_l["point"]

    return walk_merge_line(
        current_r,
        current_l,
        current_bisector,
        current_crop_point,
        go_up,
        metric,
        crossed_bisector,
        merge_array,
    )


def find_correct_w(w, nearest_neighbor, metric):
    """Ensure starting point is correct and would not result in trapped bisector"""
    starting_bisector = metric["bisector"](w, nearest_neighbor)

    w_traps = []
    for e in w["bisectors"]:
        hop_to = find_hop_to(e, w)
        is_trapped = is_bisector_trapped(hop_to, starting_bisector, metric)
        if is_trapped:
            w_traps.append({"hop_to": hop_to, "is_trapped": is_trapped})

    w_traps.sort(
        key=lambda x: metric["distance"](
            x["hop_to"]["site"], nearest_neighbor["site"]
        )
    )

    if w_traps:
        return find_correct_w(w_traps[0]["hop_to"], nearest_neighbor, metric)
    else:
        return w


def check_for_orphans(trapper, trapped, go_up, metric):
    """Check for orphaned bisectors"""
    orphans = []
    for bisector in trapped["bisectors"]:
        hop_to = find_hop_to(bisector, trapped)
        if go_up == (hop_to["site"][1] < trapped["site"][1]):
            if is_bisector_trapped(trapper, bisector, metric):
                orphans.append(bisector)

    if not orphans:
        return None

    orphans.sort(
        key=lambda a: _orphan_sort_key(a, trapped, trapper, go_up, metric)
    )
    return orphans[0] if orphans else None


def _orphan_sort_key(bisector, trapped, trapper, go_up, metric):
    """Helper for sorting orphans"""
    hop_to_a = find_hop_to(bisector, trapped)
    merge_line_a = metric["bisector"](hop_to_a, trapper)
    extreme_a = get_extreme_point(merge_line_a, go_up)

    def get_sort_val(b):
        ht = find_hop_to(b, trapped)
        ml = metric["bisector"](ht, trapper)
        return get_extreme_point(ml, go_up)

    # Simplified - just return extreme point
    return extreme_a


def determine_starting_bisector(w, nearest_neighbor, metric, last_intersect=None):
    """Determine starting bisector for the merge process"""
    z = [metric["width"], w["site"][1]]

    if not last_intersect:
        last_intersect = w["site"]

    zline = {"points": [w["site"], z]}

    intersection = None
    for bisector in nearest_neighbor["bisectors"]:
        pt = bisector_intersection(zline, bisector)
        if pt:
            intersection = {"point": pt, "bisector": bisector}
            break

    if intersection and metric["distance"](
        w["site"], intersection["point"]
    ) > metric["distance"](nearest_neighbor["site"], intersection["point"]):
        starting_bisector = metric["bisector"](w, nearest_neighbor)
        return {
            "starting_bisector": starting_bisector,
            "w": w,
            "nearest_neighbor": nearest_neighbor,
            "starting_intersection": intersection["point"]
            if intersection
            else w["site"],
        }
    elif intersection and metric["distance"](
        w["site"], intersection["point"]
    ) < metric["distance"](nearest_neighbor["site"], intersection["point"]):
        if intersection["point"][0] > last_intersect[0]:
            next_r = next(
                (e for e in intersection["bisector"]["sites"] if e != nearest_neighbor),
                nearest_neighbor,
            )
            return determine_starting_bisector(
                w, next_r, metric, intersection["point"]
            )

    w = find_correct_w(w, nearest_neighbor, metric)
    starting_bisector = metric["bisector"](w, nearest_neighbor)

    return {
        "starting_bisector": starting_bisector,
        "w": w,
        "nearest_neighbor": nearest_neighbor,
        "starting_intersection": intersection["point"] if intersection else w["site"],
    }
